extractor.go

```diff
@@ -52,30 +82,66 @@ type File struct {
 }
 
 type Extractor struct {
-	enc      *json.Encoder
-	num      int // TODO(dennwc): use it!
+	mu  sync.Mutex
+	enc *json.Encoder
+
+	num      int
 	maxSize  int64
 	symLinks bool
 }
 
+type extractJob struct {
+	fname string
+	path  string
+	buf   []byte // sample buffer
+}
+
 // Extract imports recursively from a given directory. The root is a root of the project's repository and rel is the
 // relative path inside it that will be processed. Two paths exists to allow the library to potentially parse dependency
 // manifest files that are usually located in the root of the project.
 func (e *Extractor) Extract(root, rel string) error {
+	var (
+		jobs      chan *extractJob
+		errc      chan error
+		wg        sync.WaitGroup
+		sampleBuf []byte
+	)
+	if e.num != 1 {
+		jobs = make(chan *extractJob, e.num)
+		errc = make(chan error, e.num)
+		for i := 0; i < e.num; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				e.worker(jobs, errc)
+			}()
+		}
+		// no sample buffer, it's per-routine
+	} else {
+		sampleBuf = make([]byte, e.maxSize)
+	}
 	// TODO(dennwc): expand relative imports and use dependency manifests in the future
-	return filepath.Walk(filepath.Join(root, rel), func(path string, info os.FileInfo, err error) error {
+	err := filepath.Walk(filepath.Join(root, rel), func(path string, info os.FileInfo, err error) error {
 		if err != nil {
 			return err
 		} else if info.IsDir() {
 			return nil // continue
 		}
-		sample := info.Size() > e.maxSize
 		fname, err := filepath.Rel(root, path)
 		if err != nil {
 			return err
 		}
 		if info.Mode()&os.ModeSymlink == 0 {
-			return e.processFile(fname, path, sample)
+			args := extractJob{fname: fname, path: path, buf: sampleBuf}
+			if e.num == 1 {
+				return e.processFile(args)
+			}
+			select {
+			case jobs <- &args:
+			case err = <-errc:
+				return err
+			}
+			return nil
```

Comment:
Generally, maybe we can move parallel logic to some function and just pass regular walk or parallel walk.